### PR TITLE
Refactor full frame val loss

### PR DESCRIPTION
**Pull request by maharajamihir**
- full frame prediction logic moved to train_dynamics.py. Uses sampling logic 
- migrated scan in sampling logic to nnx api
- return logits in sampling functions as well, so we can calculate metrics if needed (for validation)
- fix checkpoint loading in sampling by deleting lam decoder after initialization
- fix dynamics checkpoint loading in general by making positional encoding a simple function instead of a nnx.module or class.

### CODE REVIEW

File: train_dynamics.py [hunk 1]
```diff
@@ -479,18 +519,12 @@ def calculate_validation_metrics(val_dataloader, genie, rng):
...
-            loss_per_step.append(loss)
-            metrics_per_step.append(metrics)
-            loss_full_frame_per_step.append(loss_full_frame)
-            metrics_full_frame_per_step.append(metrics_full_frame)
+            val_outputs = val_step(genie, inputs)
+            loss_per_step.append(val_outputs["loss"])
+            metrics_per_step.append(val_outputs["metrics"])
+            if args.eval_full_frame:
+                loss_full_frame_per_step.append(val_outputs["loss_full_frame"])
+                metrics_full_frame_per_step.append(val_outputs["metrics_full_frame"])
```
[the remark this answers is not in the document]
Comment (maharajamihir) [reply train_dynamics.py:527]: no issue here.

File: train_dynamics.py [hunk 2]
```diff
@@ -596,20 +627,25 @@ def calculate_validation_metrics(val_dataloader, genie, rng):
...
-                                val_results["full_frame_seq_val"],
-                            ),
-                            axis=1,
-                        )
-                        val_results["val_full_frame_comparison_seq"] = einops.rearrange(
-                            val_results["val_full_frame_comparison_seq"] * 255,
-                            "t h w c -> h (t w) c",
-                        )
+                        if args.eval_full_frame:
+                            val_results["full_frame_seq_val"] = val_results[
```
Comment (emergenz) [inline train_dynamics.py:631]: Here, I don't think we should mutate in-place since this could lead to edge cases, e.g. when this code is executed twice consecutively without the `val_results` dict being reinitialized.

**Review by emergenz — CHANGES_REQUESTED** [review]
